probe-rs/src/probe.rs

```diff
@@ -648,6 +653,13 @@ pub trait DebugProbe: Send + fmt::Debug {
     fn get_target_voltage(&mut self) -> Result<Option<f32>, DebugProbeError> {
         Ok(None)
     }
+
+    /// Try to get a J-Link interface from the debug probe.
+    fn try_into_jlink(&mut self) -> Result<&mut jlink::JLink, DebugProbeError> {
```

Comment:
You can do something similar to https://rust.godbolt.org/z/cPzYrdY41 . The library we originaly used was called `jaylink` and I guess the name has stuck around. Would you please correct it? https://github.com/probe-rs/probe-rs/blob/98e21bc17b754a6ef2787a7819bc2f2096dbfdab/probe-rs/src/probe/jlink/mod.rs#L366

Reply:
adding `as_any`, `as_any_mut` to every impl is a bit annoying. I tried putting it as a default method in the `DebugProbe` trait but it doesn't work due to `Sized` issues.

An alternative is to reimplement `downcast` ourselves with a bit of unsafe. Requires no changes to the traits other than `+ Any`. https://play.rust-lang.org/?version=stable&mode=debug&edition=2021&gist=3f19285adeabc2e4fa1653fd6c437ea8

When [trait upcasting](https://github.com/rust-lang/rust/issues/65991) is stable, we can change the unsafe code to upcasting to `dyn Any` and then use the stdlib's `downcast`, without any changes to the trait or any public API.

Reply:
by "simple but boilerplatey" you mean `try_into_jlink`, `try_into_cmsisdap`, `try_into_stlink`...?

the problem with that is it requires probe-rs knows about all impls, so it doesn't work at all if the user is using their own out-of-tree `ProbeDriver` impl.

Reply:
Yes, exactly.

I know about that downside, but if someone really has their own driver and needs a feature, I am happy to revisit this. I feel like for now it might be a lot of work and ugly impls (using unsafe when not strictly required) when nobody actually __needs__ it.

Reply:
I honestly disagree, it's better to not introduce something we know we'll have to change in a backwards incompatible way.